p2p/src/identify/p2p_identify_reducer.rs

```diff
@@ -0,0 +1,72 @@
+use openmina_core::{bug_condition, Substate};
+use redux::ActionWithMeta;
+
+use crate::{
+    token::{IdentifyAlgorithm, StreamKind},
+    P2pNetworkConnectionMuxState, P2pNetworkKademliaAction, P2pNetworkYamuxAction, P2pState,
+};
+
+use super::P2pIdentifyAction;
+
+impl P2pState {
+    #[cfg(feature = "p2p-libp2p")]
+    pub fn identify_reducer<Action, State>(
+        mut state_context: Substate<Action, State, P2pState>,
+        action: ActionWithMeta<&P2pIdentifyAction>,
+    ) -> Result<(), String>
+    where
+        State: crate::P2pStateTrait,
+        Action: crate::P2pActionTrait<State>,
+    {
+        let (action, _meta) = action.split();
+        let p2p_state = state_context.get_substate_mut()?;
+
+        match action {
+            P2pIdentifyAction::NewRequest { addr, .. } => {
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let p2p_state: &P2pState = state.substate()?;
+                let scheduler = &p2p_state.network.scheduler;
+                let stream_id = scheduler
```

Comment:
question about these short-circuiting accessors, should these be bug conditions? if yes, I would remove the `?` at the end, perform the check manually so that `bug_condition` can be called, and then at that point yes, the error can be returned.

NOTE: I am assuming the caller is not doing any special handling with the error, if it does ignore this comment

Reply:
In most cases yes, they should be bug conditions expecally for `P2pState` but if they happen we will know, because nothing would work, since root reducer from node waits for `P2pState` to be ready before anything is action is passed there.